experimental/wazerotest/wazerotest.go

```diff
@@ -0,0 +1,605 @@
+package wazerotest
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"testing"
+
+	"github.com/tetratelabs/wazero/api"
+	"github.com/tetratelabs/wazero/experimental"
+	"github.com/tetratelabs/wazero/internal/internalapi"
+	"github.com/tetratelabs/wazero/sys"
+)
+
+const (
+	exitStatusMarker = 1 << 63
+)
+
+// Module is an implementation of the api.Module interface, it represents a
+// WebAssembly module.
+type Module struct {
+	internalapi.WazeroOnlyType
+	exitStatus uint64
+
+	// The module name that will be returned by calling the Name method.
+	ModuleName string
+
+	// The list of functions of the module. Functions with a non-empty export
+	// names will be exported by the module.
+	Functions []*Function
```

Comment:
in our other code we seem to be using `[]Function` (specifically `[]Code`) instead of `[]*Function` and access by index. should we not also do that here for consistency within our codebase?

Reply:
I had a small argument with myself about this; I landed on pointer because the function can be referenced in the `StackFrame` type, so we can write code like this:
```go
fn := &wazerotest.Function{
  ...
}

mod := &wazerotest.Module{
  Functions: []*wazerotest.Function{
    fn,
  },
}

st := []wazerotest.StackFrame{
  {Function: fn},
}
```
That being said, all the code I wrote using this package does something like this:
```go
mod := &wazerotest.Module{
  Functions: []*wazerotest.Function{
    {
      ...
    },
  },
}

st := []wazerotest.StackFrame{
  {Function: mod.Function(0)},
}
```

I also felt like API consistency within the `wazerotest` package was a bit stronger with by-pointer, so I opted for making everything pointers.

Don't hesitate to let me know if I'm missing something, I may keep it this way otherwise :)